File: src/core/road-topology.js
const ROAD_PRIORITY = Object.freeze({ street: 0, main: 1, plaza: 2 });

export const ROAD_BITS = Object.freeze({ N: 1, E: 2, S: 4, W: 8 });

const STEPS = Object.freeze({
  "0,-1": [ROAD_BITS.N, ROAD_BITS.S, "ns"],
  "1,0": [ROAD_BITS.E, ROAD_BITS.W, "ew"],
  "0,1": [ROAD_BITS.S, ROAD_BITS.N, "ns"],
  "-1,0": [ROAD_BITS.W, ROAD_BITS.E, "ew"],
});

function keyOf(x, y) { return `${x},${y}`; }

function isWater(map, point) {
  return map.terrain[point.y * map.width + point.x] === "water";
}

export function ensureRoadCell(roadMap, x, y, kind) {
  const key = keyOf(x, y);
  let record = roadMap.get(key);
  if (!record) {
    record = { kind, connections: 0, bridgeAxis: null };
    roadMap.set(key, record);
  } else if (ROAD_PRIORITY[kind] > ROAD_PRIORITY[record.kind]) {
    record.kind = kind;
  }
  return record;
}

export function bridgeAxisAt(roadMap, x, y) {
  return roadMap.get(keyOf(x, y))?.bridgeAxis ?? null;
}

export function connectRoadCells(roadMap, first, second, kind, map) {
  const step = STEPS[`${second.x - first.x},${second.y - first.y}`];
  if (!step) throw new Error(`Conexão viária não cardinal: ${keyOf(first.x, first.y)} -> ${keyOf(second.x, second.y)}`);
  const [fromBit, toBit, axis] = step;
  const firstRoad = ensureRoadCell(roadMap, first.x, first.y, kind);
  const secondRoad = ensureRoadCell(roadMap, second.x, second.y, kind);
  if (map && (isWater(map, first) || isWater(map, second))) {
    if ((isWater(map, first) && firstRoad.bridgeAxis && firstRoad.bridgeAxis !== axis)
      || (isWater(map, second) && secondRoad.bridgeAxis && secondRoad.bridgeAxis !== axis)) return false;
    if (isWater(map, first)) firstRoad.bridgeAxis = axis;
    if (isWater(map, second)) secondRoad.bridgeAxis = axis;
  }
  firstRoad.connections |= fromBit;
  secondRoad.connections |= toBit;
  return true;
}

export function addRoadPath(roadMap, path, kind, map) {
  if (!path.length) return false;
  for (const point of path) ensureRoadCell(roadMap, point.x, point.y, kind);
  for (let index = 1; index < path.length; index += 1) {
    if (!connectRoadCells(roadMap, path[index - 1], path[index], kind, map)) return false;
  }
  return true;
}

function bridgeAxis(connections) {
  const horizontal = connections & (ROAD_BITS.E | ROAD_BITS.W);
  const vertical = connections & (ROAD_BITS.N | ROAD_BITS.S);
  if (horizontal && !vertical) return "ew";
  if (vertical && !horizontal) return "ns";
  return null;
}

function collectSpan(roadByKey, start, axis, seen) {
  const step = axis === "ew" ? [1, 0] : [0, 1];
  let first = start;
  while (true) {
    const previous = roadByKey.get(keyOf(first.x - step[0], first.y - step[1]));
    if (!previous?.bridge || previous.orientation !== axis) break;
    first = previous;
  }
  const roads = [];
  let cursor = first;
  while (cursor?.bridge && cursor.orientation === axis) {
    roads.push(cursor);
    seen.add(keyOf(cursor.x, cursor.y));
    cursor = roadByKey.get(keyOf(cursor.x + step[0], cursor.y + step[1]));
  }
  return { roads, step };
}

function connectedRoadComponent(map, roads) {
  if (!roads.length) return roads;
  const byKey = new Map(roads.map((road) => [keyOf(road.x, road.y), road]));
  const preferred = map.plaza && byKey.get(keyOf(
    map.plaza.x + Math.floor(map.plaza.width / 2),
    map.plaza.y + Math.floor(map.plaza.height / 2),
  ));
  const queue = [preferred || roads.find((road) => road.kind === "plaza") || roads[0]];
  const reached = new Set([keyOf(queue[0].x, queue[0].y)]);
  const directions = [
    [ROAD_BITS.N, 0, -1], [ROAD_BITS.E, 1, 0], [ROAD_BITS.S, 0, 1], [ROAD_BITS.W, -1, 0],
  ];
  for (let cursor = 0; cursor < queue.length; cursor += 1) {
    const road = queue[cursor];
    for (const [bit, dx, dy] of directions) {
      if (!(road.connections & bit)) continue;
      const nextKey = keyOf(road.x + dx, road.y + dy);
      const next = byKey.get(nextKey);
      if (!next || reached.has(nextKey)) continue;
      reached.add(nextKey);
      queue.push(next);
    }
  }
  return roads.filter((road) => reached.has(keyOf(road.x, road.y)));
}

export function finalizeRoadTopology(map, roadMap) {
  const allRoads = [...roadMap.entries()].map(([key, record]) => {
    const [x, y] = key.split(",").map(Number);
    const bridge = map.terrain[y * map.width + x] === "water";
    return {
      x,
      y,
      kind: record.kind,
      bridge,
      orientation: bridge ? bridgeAxis(record.connections) : null,
      connections: record.connections,
      segmentIds: [],
      bridgeSpanId: null,
      bridgeRole: null,
      bridgeIndex: null,
    };
  }).sort((a, b) => a.y - b.y || a.x - b.x);
  const roads = connectedRoadComponent(map, allRoads);
  const roadByKey = new Map(roads.map((road) => [keyOf(road.x, road.y), road]));
  const roadIndex = new Map(roads.map((road, index) => [keyOf(road.x, road.y), index]));
  const seen = new Set();
  const bridgeSpans = [];
  for (const road of roads) {
    const key = keyOf(road.x, road.y);
    if (!road.bridge || !road.orientation || seen.has(key)) continue;
    const { roads: spanRoads, step } = collectSpan(roadByKey, road, road.orientation, seen);
    const id = `bridge-${bridgeSpans.length}`;
    const first = spanRoads[0];
    const last = spanRoads.at(-1);
    const entry = { x: first.x - step[0], y: first.y - step[1] };
    const exit = { x: last.x + step[0], y: last.y + step[1] };
    spanRoads.forEach((spanRoad, index) => {
      spanRoad.bridgeSpanId = id;
      spanRoad.bridgeIndex = index;
      spanRoad.bridgeRole = spanRoads.length === 1 ? "single"
        : index === 0 ? "start"
          : index === spanRoads.length - 1 ? "end"
            : index % 3 === 0 ? "post" : "middle";
    });
    bridgeSpans.push({
      id,
      axis: road.orientation,
      roadIndexes: spanRoads.map((spanRoad) => roadIndex.get(keyOf(spanRoad.x, spanRoad.y))),
      entry,
      exit,
      length: spanRoads.length,
    });
  }
  return { roads, bridgeSpans };
}

File: native/Core/RoadTopology.cs
using System.Text.Json.Nodes;
namespace Village.Core;
internal static class RoadTopology
{
    internal static readonly (int Bit,int Dx,int Dy,int Opposite,string Axis)[] Steps=[(1,0,-1,4,"ns"),(2,1,0,8,"ew"),(4,0,1,1,"ns"),(8,-1,0,2,"ew")];
    static int Priority(string kind)=>kind=="plaza"?2:kind=="main"?1:0;
    static RoadRecord Ensure(MapData m,Point p,string kind)
    {if(!m.RoadMap.TryGetValue(p,out var r))m.RoadMap[p]=r=new(kind);else if(Priority(kind)>Priority(r.Kind))r.Kind=kind;return r;}
    public static bool AddPath(MapData m,List<Point> path,string kind)
    {
        if(path.Count==0)return false;foreach(var p in path)Ensure(m,p,kind);
        for(int i=1;i<path.Count;i++)
        {
            var a=path[i-1];var b=path[i];var step=Steps.Single(s=>s.Dx==b.X-a.X&&s.Dy==b.Y-a.Y);
            var ar=Ensure(m,a,kind);var br=Ensure(m,b,kind);bool aw=m.Terrain[a.Y*m.Width+a.X]=="water",bw=m.Terrain[b.Y*m.Width+b.X]=="water";
            if(aw||bw){if((aw&&ar.BridgeAxis!=null&&ar.BridgeAxis!=step.Axis)||(bw&&br.BridgeAxis!=null&&br.BridgeAxis!=step.Axis))return false;if(aw)ar.BridgeAxis=step.Axis;if(bw)br.BridgeAxis=step.Axis;}
            ar.Connections|=step.Bit;br.Connections|=step.Opposite;
        }
        return true;
    }
    static string? Axis(int c) => (c&10)!=0&&(c&5)==0?"ew":(c&5)!=0&&(c&10)==0?"ns":null;
    public static void Finalize(MapData m)
    {
        var all=m.RoadMap.Select(pair=>J.O("x",pair.Key.X,"y",pair.Key.Y,"kind",pair.Value.Kind,"bridge",m.Terrain[pair.Key.Y*m.Width+pair.Key.X]=="water","orientation",m.Terrain[pair.Key.Y*m.Width+pair.Key.X]=="water"?Axis(pair.Value.Connections):null,"connections",pair.Value.Connections,"segmentIds",new JsonArray(),"bridgeSpanId",null,"bridgeRole",null,"bridgeIndex",null)).OrderBy(r=>J.I(r,"y")).ThenBy(r=>J.I(r,"x")).ToList();
        var by=all.ToDictionary(r=>J.Point(r));var reached=new HashSet<Point>();var queue=new List<JsonObject>();
        if(all.Count>0){queue.Add(by.GetValueOrDefault(m.Center)??all.FirstOrDefault(r=>J.S(r,"kind")=="plaza")??all[0]);reached.Add(J.Point(queue[0]));}
        for(int cursor=0;cursor<queue.Count;cursor++)foreach(var step in Steps)
        {var r=queue[cursor];if((J.I(r,"connections")&step.Bit)==0)continue;var p=new Point(J.I(r,"x")+step.Dx,J.I(r,"y")+step.Dy);if(by.TryGetValue(p,out var next)&&reached.Add(p))queue.Add(next);}
        m.Roads=all.Where(r=>reached.Contains(J.Point(r))).ToList();by=m.Roads.ToDictionary(r=>J.Point(r));
        var indexes=m.Roads.Select((r,i)=>(p:J.Point(r),i)).ToDictionary(t=>t.p,t=>t.i);var seen=new HashSet<Point>();var spans=new JsonArray();
        foreach(var road in m.Roads)
        {
            string axis=J.S(road,"orientation");if(!J.B(road,"bridge")||axis==""||seen.Contains(J.Point(road)))continue;
            int dx=axis=="ew"?1:0,dy=axis=="ns"?1:0;var first=road;
            while(by.TryGetValue(new(J.I(first,"x")-dx,J.I(first,"y")-dy),out var prev)&&J.B(prev,"bridge")&&J.S(prev,"orientation")==axis)first=prev;
            var group=new List<JsonObject>();JsonObject? current=first;
            while(current!=null&&J.B(current,"bridge")&&J.S(current,"orientation")==axis){group.Add(current);seen.Add(J.Point(current));current=by.GetValueOrDefault(new(J.I(current,"x")+dx,J.I(current,"y")+dy));}
            string id=$"bridge-{spans.Count}";var last=group[^1];
            for(int i=0;i<group.Count;i++){group[i]["bridgeSpanId"]=id;group[i]["bridgeIndex"]=i;group[i]["bridgeRole"]=group.Count==1?"single":i==0?"start":i==group.Count-1?"end":i%3==0?"post":"middle";}
            spans.Add(J.O("id",id,"axis",axis,"roadIndexes",J.A(group.Select(r=>indexes[J.Point(r)])),"entry",J.O("x",J.I(first,"x")-dx,"y",J.I(first,"y")-dy),"exit",J.O("x",J.I(last,"x")+dx,"y",J.I(last,"y")+dy),"length",group.Count));
        }
        m.BridgeSpans=spans;
    }
}
